readhomer_atlas/library/importers.py

```diff
@@ -1,70 +1,111 @@
+    """
+    urn:cts:CTSNAMESPACE:WORK:PASSAGE
+    https://cite-architecture.github.io/ctsurn_spec
+    """
 
-    book_obj = book_lookup.get(book_ref)
-    if book_obj is None:
-        book_obj, _ = Book.objects.get_or_create(
-            version=version_obj,
-            position=int(book_ref),
-            idx=counters["book_idx"],
-            ref=book_ref,
-            urn=Book.generate_urn(version_obj.urn, book_ref),
+    CTS_URN_SCHEME = ["nid", "namespace", "textgroup", "work", "version"]
+
+    def __init__(self, version_data, nodes=dict()):
+        self.version_data = version_data
+        self.nodes = nodes
+        self.urn = self.version_data["urn"].strip()
+        self.metadata = self.version_data["metadata"]
+        self.citation_scheme = self.metadata["citation_scheme"]
+        self.name = self.metadata["work_title"]
+        self.idx_lookup = defaultdict(int)
+        self.root = None
+
+    def get_node_idx(self, kind):
+        idx = self.idx_lookup[kind]
+        self.idx_lookup[kind] += 1
+        return idx
+
+    def urn_has_exemplar(self, node_urn):
+        return len(node_urn.rsplit(":")[-2].split(".")) == 4
+
+    def get_urn_scheme(self, node_urn):
+        if self.urn_has_exemplar(node_urn):
+            return [*self.CTS_URN_SCHEME, "exemplar", *self.citation_scheme]
+        return [*self.CTS_URN_SCHEME, *self.citation_scheme]
+
+    def destructure_node(self, node_urn, tokens):
+        split = ["urn:cts", ":", *re.split(r"([:|.])", node_urn)[4:]]
+        nodes = [node for idx, node in enumerate(split) if idx % 2 == 0]
+        delimiters = [delimiter for idx, delimiter in enumerate(split) if idx % 2 == 1]
+        zipped = zip(self.get_urn_scheme(node_urn), nodes)
+
+        node_data = []
+        for idx, (kind, node) in enumerate(zipped):
+            parts = nodes[: idx + 1]
+            joins = [*delimiters[:idx], ""]
+            urn = "".join(item for pair in zip(parts, joins) for item in pair)
+            if kind in self.CTS_URN_SCHEME:
+                urn = f"{urn}:"
+            data = {"kind": kind, "urn": urn}
+
+            if kind == "version":
+                data.update({"metadata": self.metadata})
+
+            if kind in self.citation_scheme:
+                ref_index = self.citation_scheme.index(kind)
+                ref = ".".join(nodes[-len(self.citation_scheme) :][: ref_index + 1])
+                data.update({"ref": ref, "rank": ref_index + 1})
+                if kind == self.citation_scheme[-1]:
+                    data.update({"text_content": tokens})
+
+            node_data.append(data)
+
+        return node_data
+
+    def generate_branch(self, line):
+        ref, tokens = line.strip().split(maxsplit=1)
+        _, passage = ref.split(".", maxsplit=1)
+        node_data = self.destructure_node(f"{self.urn}{passage}", tokens)
+        for idx, data in enumerate(node_data):
+            node = self.nodes.get(data["urn"])
+            if node is None:
```

Comment:
I've opened [Refactor CTSImporter.generate_branch to use load_bulk](https://trello.com/c/llhOXBlH) for a future sprint; this works for now